### DISCUSSION TITLE
Help with Image Processing

### DISCUSSION THREAD

**Discussion by pwberrett** [ImageMagick/ImageMagick#3977] [Help]
Hi everyone
I am new to Imagemagick. Can you please tell me whether it is possible to process images from the SOHO spacecraft in fits format in the following way?

I wish to get one days' fits images of SOHO's c2 camera, make a masterflat from those images and then divide each image by the masterflat and save the result of each image processed in this way in a separate folder as a PNG file. 

Is this possible using the command line?

Thanks
Peter

**Comment by snibgo** [comment]
> Is this possible using the command line?

I expect so. What version of IM, on what platform?

Please link to a sample input fits file, and an explanation of "masterflat".

**Comment by fmw42** [comment]
If I understand the process, then ImageMagick can do that with some OS scripting. You would read all the images (IM can read FITS format). Then I suppose average them. Then use the average to divide into each image. The command line can do the average using -evaluate-sequence average and can do the divide using -compose divide -composite.

However, I will defer to one of the photography experts on this forum for a more detailed answer.

**Reply by pwberrett** [reply]
Here is a link to my sample files. Look in the directory #1 sample fits files....

https://drive.google.com/drive/folders/1uCNJD3jNA66AQJdRdd0LszIXMo-POQ4L?ths=true

The process as I understand is not to average the files but to take the median of each pixel location. Each original fits image is then divided by the masterflat and the results of these divisions are saved into a new folder. 

so 

masterflat=median of all fits files

newfile=oldfile/masterflat
newfile2=oldfile2/masterflat 
etc etc etc

Ideally the resultant files should be adjusted for a constant brightness. 

At the end of the day I want to bulk process images in python but the easiest way seems to do it via the command line using the os library.

I am using Windows and Imagemagick 7.0.11-6 Q16 x64 2021.04.03

Explanation of masterflat images

https://www.skyatnightmagazine.com/astrophotography/astrophoto-tips/using-flat-frames-calibrate-astrophotos/

cheers Peter

**Comment by snibgo** [comment]
To find the median of the inputs (Windows BAT command):
```
magick ^
  22196759.fts ^
  22196760.fts ^
  22196761.fts ^
  22196800.fts ^
  -evaluate-sequence Median ^
  flat.png
```
I have used just four inputs. You could instead use 2219*.fts if you want.

flat.png is easily viewable. For performance, use flat.miff instead. ".miff" is an internal ImageMagick format, and is faster than png.

To divide one image by the median:
```
magick ^
  22196759.fts ^
  flat.png ^
  -compose DivideSrc -composite ^
  div0.png
```
Note that where a fts pixel has a value greater than the corresponding value in the flat, the result will be more than 100% of QuantumRange, eg "whiter than white", and will be clipped in the output png. Instead, you can save the result in floating-point tiff, which won't clip:
```
magick ^
  22196759.fts ^
  flat.png ^
  -compose DivideSrc -composite ^
  -define quantum:format=floating-point -depth 32 ^
  div0.tiff
```

EDIT: The commands could be combined into one, with no need to record the image flat.png. But I suspect the set of images used to create the flat image are a different set to those that are divided by the flat image.
